### PR TITLE
Final prototype comments and top level statements local adjustments

### PR DESCRIPTION
**Pull request by 333fred**
Clean up the last of the prototype comments and adjust the parsing of locals named required in top level statements.

Test plan https://github.com/dotnet/roslyn/issues/57046

Fixes https://github.com/dotnet/roslyn/issues/61510.

### CODE REVIEW

**Comment by 333fred** [comment]
@jcouv @RikkiGibson for review.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -1258,8 +1258,9 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
                     case DeclarationModifiers.Required:
                         // In C# 11, required in a modifier position is always a keyword if not escaped. Otherwise, we reuse the async detection
                         // machinery to make a conservative guess as to whether the user meant required to be a keyword, so that they get a good langver
-                        // diagnostic and all the machinery to upgrade their project kicks in.
-                        if (!IsFeatureEnabled(MessageID.IDS_FeatureRequiredMembers) && !ShouldAsyncOrRequiredBeTreatedAsModifier(parsingStatementNotDeclaration: false))
+                        // diagnostic and all the machinery to upgrade their project kicks in. The only exception to this rule is top level statements,
+                        // where the user could conceivably have a local named required. For these locations, we need to disambiguate as well.
+                        if ((!IsFeatureEnabled(MessageID.IDS_FeatureRequiredMembers) || forTopLevelStatements) && !ShouldAsyncOrRequiredBeTreatedAsModifier(parsingStatementNotDeclaration: false))
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1263]: This is the meat of the top-level-statement parsing change. Previously, we'd _always_ treat `required` as a modifier for global code in C# 11, and `required = ...;` as a top-level statement fits that category. We now fall back to disambiguation for this case.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1263]: Fixes [https://github.com/dotnet/roslyn/issues/61510](https://github.com/dotnet/roslyn/issues/61510) ?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1263]: Yes.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -1173,7 +1173,7 @@ internal static DeclarationModifiers GetModifier(SyntaxKind kind, SyntaxKind con
             }
         }
 
-        private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
+        private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors, bool forTopLevelStatements)
         {
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1177]: nit: should we `Debug.Assert(!(forAccessors && forTopLevelStatements))`?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1263]: To be totally clear--if we wanted `async`-like behavior where types can be named `async`, we would be able to do something like the `DeclarationModifiers.Async` section above, right?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1263]: Yes, we'd do something like that, at the cost of making the parse trees for these scenarios worse.

**Pull request merged**